crypto/ring-signature/signer/src/traits.rs

```diff
@@ -73,6 +73,10 @@ impl From<RistrettoPrivate> for OneTimeKeyDeriveData {
 ///
 /// A transaction builder can be built around this.
 pub trait RingSigner {
+    /// Implementation errors must be convertable to standard [SignerError] type
+    /// for common handling
+    type Error: Into<SignerError>;
```

Comment:
So, what is the advantage of using a generic associated type for `Error` here?

It seems that, with this trait bound, the only thing a caller to `RingSigner` can possibly do with this is convert the given error to `SignerError`. And the callee must implement a way to do that conversion. So why not cut to the chase and always return `SignerError`? It seems like we just have more boilerplate and more conversions pushed onto the caller if we go this route.

Reply:
It also makes `RingSigner` not object safe, so we cannot have like `Box<dyn RingSigner>` which may be useful.

Reply:
I could understand if you were trying to like, break a dependency cycle or something. But it seems that, any implementer of RingSigner must depend on this crate, for this trait, and for this error type now -- they still have to have a definition of `SignerError` to implement `Into<SignerError>`. So I can't understand what we're getting by doing this.

Reply:
the concept was that it should be possible to pass up things like transport errors for the ring signer implementation,  for the moment they just get squashed in the transaction signer anyway, but ideally they'd be propagated out.

for an object safe alternate we could have an `Transport<Box<dyn Error>>` so the error type is fixed but we can include transport-specific information?

Reply:
So I think what I'd suggest is, let's try to think about who is using the `RingSigner`, and how we think they might handle transport errors. Presumably there's like, retriable transport errors, and fatal transport errors? So maybe we can add some possibilities like that in `SignerError`? And otherwise the transport error details are just logged?

And if we find we need more categories later, we can simply add more possibilities to `SignerError`?

Reply:
it just seems like that's the simplest thing that would work -- the caller probably doesn't need all the details of the transport error, just how they should handle it.

and if we're going to have a "common error" type that all implementations work with, for "common handling", why not just make that the error type.

clearly, we're expecting that the callers are going to squash the implementation specific errors this way anyways, right? so when you say this:

>  but ideally they'd be propagated up (and the error type is consistent with other interfaces, not just the ring signer).

you're not actually coding towards that, right? you're sort of giving the implementation the option to return whatever error they want, but the caller is then going to have to squash it, whatever it is

Reply:
i'm still thinking about this part:

>  (and the error type is consistent with other interfaces, not just the ring signer)

can you give a concrete example where that is really helpful?

Reply:
it's probably not helpful, just, consistent. i guess the opposite question is, should we define concrete error types for [other traits](https://github.com/mobilecoinfoundation/mobilecoin/blob/9fbee911bcd183d8afb042d5c9e07bbac7185915/transaction/signer/src/traits.rs#L20)?

i think i actually did look at embedding a boxed type, but there's a serde constraint on [`SignerError`](https://github.com/mobilecoinfoundation/mobilecoin/blob/master/crypto/ring-signature/signer/src/traits.rs#L127) which wasn't so straightforward to work around

Reply:
Hmm, I wonder why we put `serde` on the `SignerError`, that's somewhat unusual

I'm okay with putting concrete errors for the other traits? I guess they should all accommodate transport error

If you want to go the other way, we could define an additional trait for these errors, and make all of them implement it? We did something like that here:

https://github.com/mobilecoinfoundation/mobilecoin/blob/f106b4fa8a03779ddf287b041687685d386db843/connection/src/traits.rs#L31

Reply:
I would be okay with generic errors for `RingSigner` and a way to interrogate them as well, and make the current `SignerError` only be the error type for the `NoKeysRingSigner` etc. I just think if we're going to make all errors convertible to `SignerError`, then that should be the error type always.

Reply:
the conversion bounds trick is just one way to support interrogating generic errors, but it's nbd either way.  have reverted the changes to the error for now / will introduce more states as required ^_^